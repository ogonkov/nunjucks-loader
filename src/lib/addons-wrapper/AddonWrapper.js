import {stringifyRequest} from 'loader-utils';

import {getModule} from '../../public/utils/get-module';
import {IMPORTS_PREFIX, TEMPLATE_DEPENDENCIES} from '../constants';
import {getModuleOutput} from '../output/get-module-output';
import {getImportStr} from '../utils/get-import-str';
import {toVar} from '../utils/to-var';


export class AddonWrapper {
    #name = null;
    #importPath = null;
    #instance = null;
    #type = null;
    #loaderContext = null;
    #es = false;

    constructor({name, importPath, type, es, loaderContext}) {
        if (typeof name !== 'string') {
            throw new TypeError('AddonWrapper: name should be a string');
        }
        this.#name = name;

        if (!importPath) {
            throw new TypeError('AddonWrapper: import path required');
        }
        this.#importPath = importPath;

        if (typeof type !== 'string') {
            throw new TypeError('AddonWrapper: addon type required');
        }
        this.#type = type;

        this.#loaderContext = loaderContext;

        if (typeof es !== 'boolean') {
            throw new TypeError('AddonWrapper: es type should be a boolean');
        }
        this.#es = es === true;
    }

    get name() {
        return this.#name;
    }

    get instance() {
        if (this.#instance !== null) {
            return this.#instance;
        }

        return new Promise((resolve, reject) => {
            import(this.#importPath)
                .then((instance) => getModule(instance))
                .then((_instance) => {
                    this.#instance = _instance;
                    return _instance;
                })
                .then(resolve)
                .catch(reject);
        });
    }

    get importVar() {
        return toVar(`${IMPORTS_PREFIX}_${this.#type}_${this.#name}`);
    }

    get importStatement() {
        return getImportStr(
            stringifyRequest(this.#loaderContext, this.#importPath),
            this.#es
        )(this.importVar);
    }

    get dependencyInject() {
        return `${TEMPLATE_DEPENDENCIES}.${this.#type}['${this.#name}'] = {
            module: ${getModuleOutput(this.importVar)}
        };`;
    }
}
