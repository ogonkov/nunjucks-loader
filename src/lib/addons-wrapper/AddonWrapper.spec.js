import {AddonWrapper} from './AddonWrapper';

describe('constructor', function() {
    it('should throw with wrong name', function() {
        expect(() => new AddonWrapper({})).toThrowErrorMatchingSnapshot();
    });

    it('should throw w/out import path', function() {
        expect(() => new AddonWrapper({
            name: 'foo'
        })).toThrowErrorMatchingSnapshot();
    });

    it('should throw w/out type', function() {
        expect(() => new AddonWrapper({
            name: 'foo',
            importPath: 'bar'
        })).toThrowErrorMatchingSnapshot();
    });

    it('should throw with wrong es flag', function() {
        expect(() => new AddonWrapper({
            name: 'foo',
            importPath: 'bar',
            type: 'x'
        })).toThrowErrorMatchingSnapshot();
    });
});
