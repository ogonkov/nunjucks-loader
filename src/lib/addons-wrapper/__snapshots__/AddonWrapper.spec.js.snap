// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor should throw w/out import path 1`] = `"AddonWrapper: import path required"`;

exports[`constructor should throw w/out type 1`] = `"AddonWrapper: addon type required"`;

exports[`constructor should throw with wrong es flag 1`] = `"AddonWrapper: es type should be a boolean"`;

exports[`constructor should throw with wrong name 1`] = `"AddonWrapper: name should be a string"`;
